### PR TITLE
Abocquet/votes frontend

### PR DESCRIPTION
**Pull request by abocquet**

### CODE REVIEW

File: backend/associations/serializers/election_read_only.py [hunk 1]
```diff
@@ -18,14 +17,16 @@ def to_representation(self, instance: Choice):
 
         if instance.show_results:
             response["number_of_votes"] = instance.number_of_votes
+            response["number_of_offline_votes"] = instance.number_of_offline_votes
+            response["number_of_online_votes"] = instance.number_of_online_votes
```
Comment (Piwakk) [inline backend/associations/serializers/election_read_only.py:21]: Why do we need this? In my opinion, this information is on a need to know basis

File: frontend/src/components/associations/elections/edit/Administration.tsx [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+import { api } from "../../../../services/apiService";
+import { ToastContext } from "../../../utils/Toast";
+
+export const Administration = ({ election }: { election: Election }) => {
+  const { sendInfoToast, sendSuccessToast, sendErrorToast } = useContext(
+    ToastContext
+  );
+
+  const [save] = useMutation(api.elections.update, {
+    onMutate: () => sendInfoToast("Sauvergarde..."),
```
Comment (Piwakk) [inline frontend/src/components/associations/elections/edit/Administration.tsx:18]:
```suggestion
    onMutate: () => sendInfoToast("Sauvegarde..."),
```

File: frontend/src/components/associations/elections/edit/Administration.tsx [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    onSuccess: () => {
+      sendSuccessToast("Sauvegardée !");
+      queryCache.invalidateQueries("election.get");
+    },
+    onError: (err) =>
+      sendErrorToast(`Erreur lors de la sauvegarde: ${err.toString}`),
```
Comment (Piwakk) [inline frontend/src/components/associations/elections/edit/Administration.tsx:24]:
```suggestion
      sendErrorToast(`Erreur lors de la sauvegarde : ${err.toString}`),
```

File: frontend/src/components/associations/elections/edit/Administration.tsx [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+      <Form>
+        <Card>
+          <Card.Header>
+            <Card.Title>Paramètres</Card.Title>
+          </Card.Header>
+          <Card.Body>
+            <TextFormGroup
+              name={"name"}
+              label={"Intitulé de l'élection"}
+              placeholder={"Nouvelle éléction"}
```
Comment (Piwakk) [inline frontend/src/components/associations/elections/edit/Administration.tsx:42]:
```suggestion
              placeholder={"Nouvelle élection"}
```

File: frontend/src/components/associations/elections/edit/RegistrationList.tsx [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import { ToastContext } from "../../../utils/Toast";
+
+export const RegistrationList = ({ election }) => {
+  const { sendSuccessToast, sendErrorToast } = useContext(ToastContext);
+  const [initialized, setInitialized] = useState(false);
+
+  const sendToast = (promise) => {
+    promise
+      .then(() => {
+        sendSuccessToast("Liste des electeurs sauvegardée");
```
Comment (Piwakk) [inline frontend/src/components/associations/elections/edit/RegistrationList.tsx:13]:
```suggestion
        sendSuccessToast("Liste des électeurs sauvegardée");
```

File: frontend/src/components/associations/elections/edit/RegistrationList.tsx [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+      })
+      .catch((err) => {
+        sendErrorToast(
+          "Erreur lors de la sauvegarde de la liste des electeurs " +
```
Comment (Piwakk) [inline frontend/src/components/associations/elections/edit/RegistrationList.tsx:17]:
```suggestion
          "Erreur lors de la sauvegarde de la liste des électeurs " +
```

File: frontend/src/components/associations/elections/view/Vote.tsx [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    );
+  }, [election.choices]);
+
+  const submitVote = () => {
+    if (
+      checks.filter(({ choice, state }) => state).length >
+      election.maxChoicesPerBallot
+    ) {
+      sendErrorToast(
+        `Sélctionnez ${election.maxChoicesPerBallot} choix maximum`
```
Comment (Piwakk) [inline frontend/src/components/associations/elections/view/Vote.tsx:27]:
```suggestion
        `Sélectionnez ${election.maxChoicesPerBallot} choix maximum`
```

File: frontend/src/components/utils/ErrorPage.tsx [hunk 8]
```diff
@@ -7,16 +7,24 @@ export const ErrorMessage = (props) => {
...
-export const ErrorPage = (props) => {
+export const ErrorPage = ({
+  errorCode = 500,
+  title = "",
+  children = null,
+}: {
+  errorCode?: number;
+  title?: string;
+  children?: JSX.Element | string | Error | null;
+}) => {
```
Comment (Piwakk) [inline frontend/src/components/utils/ErrorPage.tsx:18]: 😲

File: frontend/src/components/utils/UserSelector.tsx [hunk 9]
```diff
@@ -1,22 +1,42 @@
...
+  getUser = (user) => user,
+}: {
+  setUser: (user) => void;
+  title: string;
+  helper?: string;
+  inCard?: boolean;
+  queryParams?: object;
+  apiKey?: string;
+  apiMethod?: any;
+  getUser?: (user) => User;
```
Comment (Piwakk) [inline frontend/src/components/utils/UserSelector.tsx:28]: 😲 😲

File: frontend/src/models/associations/election.ts [hunk 10]
```diff
@@ -2,23 +2,38 @@ import { User } from "../user";
 import { Association } from "./association";
 
 export interface Choice {
-  id: string;
-  election: Election;
+  id: number;
+  name: string;
+  numberOfOfflineVotes: number;
+  numberOfOnlineVotes: number;
```
Comment (Piwakk) [inline frontend/src/models/associations/election.ts:8]: The camel case of `offline` and `online` is not consistent with frontend/src/components/associations/elections/view/View.tsx

File: frontend/src/models/associations/election.ts [hunk 11]
```diff
@@ -2,23 +2,38 @@ import { User } from "../user";
...
-export interface Ballot {
-  id: string;
-  election: Election;
-  choices: Choice[];
+type VoteStatus = "PENDING" | "OFFLINE_VOTE" | "ONLINE_VOTE";
+export interface UserVoter {
+  id: number;
+  user: User;
+  status: VoteStatus;
+  election: 5;
```
Comment (Piwakk) [inline frontend/src/models/associations/election.ts:17]: ?

File: backend/associations/serializers/election_read_only.py (hunk 1 above)
Comment (abocquet) [reply backend/associations/serializers/election_read_only.py:21]: This is useful to give the detail of the votes. This is the way it is done today.
image

**Pull request merged**